modules/Logger.js

```diff
@@ -0,0 +1,169 @@
+              }
+          }
+        }
+      }
+  	},
+
+    /**
+    * tracks information about the 'start' of events
+    * @param {JSON object} contains tracking information including UUID and time
+    * @method startEvent
+    */
+    startEvent: function(eventInfo){
+      var uuid = eventInfo.UUID;
+      eventInfo.sessionID = this._sessionID; //check if event info has sessionID property and throw error
+      eventInfo.startTime = eventInfo.time;
+      delete eventInfo.time;
+      this.unclosed[uuid] = eventInfo;
+    },
+
+    /**
+    * tracks information about the 'end' of events
+    * @param {JSON object} contains information about the tracked event, including UUID, time, state
+    * @method endEvent
+    */
+    endEvent: function(eventInfo){
+      var uuid = eventInfo.UUID;
+      var totalEvent = this.unclosed[uuid];
+      var duration = eventInfo.time - totalEvent.startTime; 
+      totalEvent.duration = duration;
+      totalEvent.endTime = eventInfo.time;
+      this._sessionEnd = eventInfo.time;
+      delete totalEvent.state;
+      this.log[uuid] = totalEvent;
+      delete this.unclosed[uuid];
+    },
+
+    /**
+    * tracks information about the beginning of sessions
+    * @param {JSON object} contains information about the tracked event
+    * @method openSession
+    */
+    openSession: function(eventInfo){
+      this._isClosed = false;
+      var sessionID = (new Date()).getTime().toString(16)+Math.floor(1E7*Math.random()).toString(16);
+      this._sessionID = sessionID;
+      this._sessionStart = eventInfo.time;
+    },
+
+    /**
+    * tracks information about the end of a session. resets all relevant parameters
+    * @method closeSession
+    */
+    closeSession: function(){
+      var sessionEvent = {sessionID: this._sessionID, type: "session"};
+      var sessionDuration = this._sessionEnd - this._sessionStart;
+      sessionEvent.sessionDuration = sessionDuration;
+      this.sessions.push(sessionEvent);
+      for (var key in this.log){
+        this.log[key].sessionDuration = sessionDuration;
+        this.sessions.push(this.log[key]);
+        delete this.log[key];
+      }
+      this._isClosed = true;
+      this.sendToNewRelic();
+      this.resetSession();
+    },
+
+    /**
+    * reset all relevant parameters to prepare for new session
+    * @method resetSession
+    */
+    resetSession: function(){
+      this.log = {};
+      this.unclosed = {};
+      this.sessions = [];
+      this._isClosed = true;
+      this._sessionStart = null;
+      this._sessionEnd = null;
+      this._sessionID = null;
+    }, 
+
+    /**
+    * send all information from the session to new relic
+    * @method sendToNewRelic
+    */
+    sendToNewRelic: function(){
+      var sessionLength = this.sessions.length;
+      for (var i = 0; i < sessionLength; i++){
+        var evt = this.sessions[i];
+        var actionName = evt.type;
+        newrelic.addPageAction(actionName, evt);
```

Comment:
there is a hidden dependency in this module on a variable "newrelic" in global scope. This needs to be explicit dependency and also optional for use with torso. We should be checking if window.newrelic exists and throwing an exception in this method if it doesn't.